src/requirements/requirement-graph.ts

```diff
@@ -0,0 +1,91 @@
+import { HashMap, HashSet } from './util/collections';
+
+/**
+ * A mutable graph data structure that represents the mathematically relation between requirement
+ * and course.
+ * An edge between requirement `r` and course `c` implies that `c` may help satisfy `r`. It's up to
+ * the client of the data structure to gradually refine the graph so that an edge `r-c` implies that
+ * `c` definitely helps satisfies `r`.
```

Comment:
so would it continue after 'r-c' to 'r-c-r'? How does the relationship between 'c-r' work and what does that mean?

Reply:
No, the graph is undirected. So `r-c` automatically implies `c-r`